### DISCUSSION TITLE
Use ipython as an internal shell?

### DISCUSSION THREAD

**Discussion by jgarte** [inducer/pudb#618] [Q&A] [answered]
Hi,

Is there a way to use ipython as an internal shell. That is, when pressing `ctrl x`?

I was able to use the ipython REPL as an external shell when pressing `!` but not `ctrl x`.

Would this turn into a new feature request to integrate ipython as an internal shell?

What are the current difficulties with supporting ipython as an internal shell? I am not aware of them yet.

**Comment by inducer** [comment]
When in the internal shell, Urwid is managing interactions with the terminal. (I.e. the internal shell is not a general-purpose terminal.) That's needed so that the output stays contained in the appropriate subwindow. So if you can get IPython to do its I/O through Urwid, I think there's a chance. But it would still be a tricky engineering problem.

**Reply by cohml** [reply]
Can you provide some links to resources where one could look into this? Thanks in advance, @inducer.

I've never heard of Urwid, nor do I have any idea what "managing interactions with the terminal" might consist of. But I absolutely hate the vanilla Python REPL, and absolutely love iPython, so I'd like to consider putting in the work.

Being able to use iPython as an internal shell - not having to leave `pudb` to get a terminal; being able to use vi mode - would be an absolute game-changer.

**Reply by inducer** [reply, marked as answer]
Think of Urwid as something that allows its user to think about nice abstract things like "Window" and "text entry box" and "key event" and then talks the required ANSI escape sequence gobbledygook to the terminal.

The problem here is that both Urwid and IPython by default talk ANSI escape sequences to the terminal. Without some sort of integration, IPython will assume it owns the whole screen and trample over the output lovingly created by Urwid. IPython *can* be taught to respect other UI toolkits, as evidenced by `ptipython` (that's prompt-toolkit-ipython). I don't know exactly how much work is involved to do that, but that might be a good place to look for an example of how to teach ipython to talk to different UI toolkits.
